cpp/include/cudf/ast/detail/expression_evaluator.cuh

```diff
@@ -70,14 +73,15 @@ struct expression_result {
   // used, whereas passing it as a parameter keeps it in registers for fast
   // access at the point where indexing occurs.
   template <typename Element>
-  __device__ void set_value(cudf::size_type index, possibly_null_value_t<Element, has_nulls> result)
+  CUDA_DEVICE_CALLABLE void set_value(cudf::size_type index,
+                                      possibly_null_value_t<Element, has_nulls> const& result)
   {
-    subclass()->set_value();
+    subclass().template set_value<Element>(index, result);
   }
 
-  __device__ bool is_valid() const { subclass()->is_valid(); }
+  CUDA_DEVICE_CALLABLE bool is_valid() const { return subclass().is_valid(); }
 
-  __device__ T value() const { subclass()->value(); }
+  CUDA_DEVICE_CALLABLE T value() const { return subclass()->value(); }
```

Comment:
Just curious about this line since `subclass()` does not return a pointer.

Reply:
Good catch. If you'll notice above I fixed this problem in the `is_valid` function but apparently I missed it in `value`. Prior to this PR I had converted the class structure of the results to use CRTP, but I forgot to update the templating in the methods of the `expression_evaluator` so the call sites always knew what subclass of `expression_result` they were receiving. As a result, the parent `expression_output` methods were never being called. Now that the templating has been fixed to enforce the desired API, the code inside the `expression_evaluator` is in fact calling the appropriate parent methods (which then dispatch via `subclass()`), but only `is_valid` and `set_value` are ever called inside the evaluation.

The `value` method is how consumers of this API retrieve outputs. In particular, the conditional join code evaluates the predicate to determine whether a pair of rows joins or not. However, at _that_ call site the type of the `expression_result` subclass is also known, so again the subclass's method is automatically called. All of this is to say that `expression_result.value` is never actually called, which is why I'm guessing this never threw any compile- or run-time errors. Still good to fix now though.